### PR TITLE
dubbo_proxy: Add filter extension mechanism for Dubbo protocol

### PR DESCRIPTION
**Pull request by gengleilei**
*Description*: Add filter extension mechanism for Dubbo protocol
*Risk Level*: low
*Testing*: N/A
*Docs Changes*: N/A
*Release Notes*: N/A
[Optional Fixes #Issue] #3998 
[Optional *Deprecated*:]

this is only part of the code for DubboProxy, the rest of the code hasn't been committed yet, complete implementation: https://github.com/gengleilei/envoy/tree/feature-dubbo-router-develop/source/extensions/filters/network/dubbo_proxy

### CODE REVIEW

File: source/extensions/filters/network/dubbo_proxy/decoder_event_handler.h [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+  /**
+   * Transfer the original header data of the Dubbo protocol,
+   * it's called after the protocol's header data is parsed.
+   * @param header_buf raw header data
+   * @param size The size of the head.
+   */
+  virtual Network::FilterStatus transferHeaderTo(const Buffer::Instance& header_buf, size_t size) {
+    if (buffer_ != nullptr) {
+      Buffer::OwnedImpl copy(header_buf);
```
Comment (zyfjeff) [inline source/extensions/filters/network/dubbo_proxy/decoder_event_handler.h]: Why are you copying here?

File: source/extensions/filters/network/dubbo_proxy/decoder_event_handler.h [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+  /**
+   * Transfer the original body data of the Dubbo protocol
+   * it's called after the protocol's body data is parsed.
+   * @param header_buf raw body data
+   * @param size The size of the body.
+   */
+  virtual Network::FilterStatus transferBodyTo(const Buffer::Instance& body_buf, size_t size) {
+    if (buffer_ != nullptr) {
+      Buffer::OwnedImpl copy(body_buf);
```
Comment (zyfjeff) [inline source/extensions/filters/network/dubbo_proxy/decoder_event_handler.h]: ditto

File: source/extensions/filters/network/dubbo_proxy/decoder_event_handler.h [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+  }
+
+  /**
+   * Transfer the original body data of the Dubbo protocol
+   * it's called after the protocol's body data is parsed.
+   * @param header_buf raw body data
+   * @param size The size of the body.
+   */
+  virtual Network::FilterStatus transferBodyTo(const Buffer::Instance& body_buf, size_t size) {
+    if (buffer_ != nullptr) {
```
Comment (zyfjeff) [inline source/extensions/filters/network/dubbo_proxy/decoder_event_handler.h]: ASSERT(buffer_)?

File: source/extensions/filters/network/dubbo_proxy/router/router.h [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+  /**
+   * @return const std::string& the upstream cluster that owns the route.
+   */
+  virtual const std::string& clusterName() const PURE;
+
+  /**
+   * @return MetadataMatchCriteria* the metadata that a subset load balancer should match when
+   * selecting an upstream host
+   */
+  virtual const Envoy::Router::MetadataMatchCriteria* metadataMatchCriteria() const PURE;
```
Comment (zyfjeff) [inline source/extensions/filters/network/dubbo_proxy/router/router.h]: `Envoy::Router::MetadataMatchCriteria`   forward statement?

File: source/extensions/filters/network/dubbo_proxy/router/router.h [hunk 5]
```diff
@@ -0,0 +1,74 @@
+#pragma once
+
+#include <memory>
+#include <string>
+
+#include "envoy/router/router.h"
```
Comment (zyfjeff) [inline source/extensions/filters/network/dubbo_proxy/router/router.h]: Remove it. use forward statement?

File: source/extensions/filters/network/dubbo_proxy/router/BUILD [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+    "envoy_cc_library",
+    "envoy_package",
+)
+
+envoy_package()
+
+envoy_cc_library(
+    name = "router_interface",
+    hdrs = ["router.h"],
+    external_deps = ["abseil_optional"],
```
Comment (zyfjeff) [inline source/extensions/filters/network/dubbo_proxy/router/BUILD]: Remove it?

File: source/extensions/filters/network/dubbo_proxy/decoder_event_handler.h (hunk 1 above)
Comment (gengleilei) [reply source/extensions/filters/network/dubbo_proxy/decoder_event_handler.h]: Users can define multiple Dubbo filters, and some filters will not get data if we use move.

File: source/extensions/filters/network/dubbo_proxy/decoder_event_handler.h (hunk 2 above)
Comment (gengleilei) [reply source/extensions/filters/network/dubbo_proxy/decoder_event_handler.h]: Users can define multiple Dubbo filters, and some filters will not get data if we use move.

File: source/extensions/filters/network/dubbo_proxy/decoder_event_handler.h (hunk 3 above)
Comment (gengleilei) [reply source/extensions/filters/network/dubbo_proxy/decoder_event_handler.h]: I hope ProtocolDataPassthroughConverter interface is optional, but if you use ASSERT means that the user must use the interface.

File: source/extensions/filters/network/dubbo_proxy/router/router.h (hunk 4 above)
Comment (gengleilei) [reply source/extensions/filters/network/dubbo_proxy/router/router.h]: I don't think it is appropriate to use the forward statement here. the router interface is provided to the user, and I want to express the dependency explicitly. If this is a implement class, it is best to use the forward statement.

File: source/extensions/filters/network/dubbo_proxy/router/router.h (hunk 5 above)
Comment (gengleilei) [reply source/extensions/filters/network/dubbo_proxy/router/router.h]: ditto

File: source/extensions/filters/network/dubbo_proxy/router/BUILD (hunk 6 above)
Comment (gengleilei) [reply source/extensions/filters/network/dubbo_proxy/router/BUILD]: done

**Comment by lizan** [comment]
@zyfjeff Sorry I'm busy this week, ping me once this PR looks good to you.

File: source/extensions/filters/network/dubbo_proxy/filters/well_known_names.h [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+namespace Extensions {
+namespace NetworkFilters {
+namespace DubboProxy {
+namespace DubboFilters {
+
+/**
+ * Well-known http filter names.
+ * NOTE: New filters should use the well known name: envoy.filters.dubbo.name.
+ */
+class DubboFilterNameValues {
```
Comment (zyfjeff) [inline source/extensions/filters/network/dubbo_proxy/filters/well_known_names.h]: struct?

**Comment by zyfjeff** [comment]
@lizan  When you have time, you can take a look at this PR. it looks good for me

File: source/extensions/filters/network/dubbo_proxy/filters/well_known_names.h (hunk 7 above)
Comment (gengleilei) [reply source/extensions/filters/network/dubbo_proxy/filters/well_known_names.h]: class is used to keep the style consistent with other NameValues definitions

File: source/extensions/filters/network/dubbo_proxy/decoder_event_handler.h [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+
+/**
+ * This class provides the pass-through capability of the original data of
+ * the Dubbo protocol to improve the forwarding efficiency
+ * when no modification of the original data is required.
+ * Note: If the custom filter does not care about data transfer,
+ *       then it does not need to care about this interface,
+ *       which is currently used by router filter.
+ */
+class ProtocolDataPassthroughConverter {
```
[the remark this answers is not in the document]
Comment (gengleilei) [reply source/extensions/filters/network/dubbo_proxy/decoder_event_handler.h]: when downstream's source data did not need to be modified, we hope that this interface was used when downstream was directly transmitted to upstream. decoder would call transferXXXTo interface transfer the source data，which is currently used by Dubbo router filter.

File: test/extensions/filters/network/dubbo_proxy/mocks.cc [hunk 9]
```diff
@@ -1,20 +1,97 @@
...
+  ON_CALL(*this, streamInfo()).WillByDefault(ReturnRef(stream_info_));
+}
+MockDecoderFilterCallbacks::~MockDecoderFilterCallbacks() {}
+
+MockDirectResponse::MockDirectResponse() {}
+MockDirectResponse::~MockDirectResponse() {}
+
+MockFilterConfigFactory::MockFilterConfigFactory()
+    : MockFactoryBase("envoy.filters.dubbo.mock_filter") {
+  mock_filter_.reset(new NiceMock<MockDecoderFilter>());
```
[the remark this answers is not in the document]
Comment (gengleilei) [reply test/extensions/filters/network/dubbo_proxy/mocks.cc]: done

File: test/extensions/filters/network/dubbo_proxy/mocks.h [hunk 10]
```diff
@@ -1,16 +1,37 @@
...
+
 namespace Envoy {
 namespace Extensions {
 namespace NetworkFilters {
 namespace DubboProxy {
 
+class MockDecoderEventHandler : public DecoderEventHandler {
+public:
+  MockDecoderEventHandler();
+  virtual ~MockDecoderEventHandler() override {}
```
[the remark this answers is not in the document]
Comment (gengleilei) [reply test/extensions/filters/network/dubbo_proxy/mocks.h]: done

File: test/extensions/filters/network/dubbo_proxy/mocks.cc [hunk 11]
```diff
@@ -1,20 +1,97 @@
...
+}
+
+MockFilterConfigFactory::~MockFilterConfigFactory() {}
+
+FilterFactoryCb MockFilterConfigFactory::createFilterFactoryFromProtoTyped(
+    const ProtobufWkt::Struct& proto_config, const std::string& stat_prefix,
+    Server::Configuration::FactoryContext& context) {
```
[the remark this answers is not in the document]
Comment (gengleilei) [reply test/extensions/filters/network/dubbo_proxy/mocks.cc]: done

**Comment by gengleilei** [comment]
@lizan Is there any other code review opinion ?

**Pull request merged**